packages/3dom/src/api.ts

```diff
@@ -0,0 +1,199 @@
+      options?: boolean|AddEventListenerOptions): void;
+  addEventListener(
+      type: string, listener: EventListenerOrEventListenerObject,
+      options?: boolean|AddEventListenerOptions): void;
+
+  removeEventListener<K extends keyof ThreeDOMEventMap>(
+      type: K,
+      listener: (this: ThreeDOMGlobalScope, ev: ThreeDOMEventMap[K]) => unknown,
+      options?: boolean|EventListenerOptions): void;
+  removeEventListener(
+      type: string, listener: EventListenerOrEventListenerObject,
+      options?: boolean|EventListenerOptions): void;
+
+  /**
+   * A reference to Model constructor. Supports instanceof checks; this class is
+   * not directly constructable.
+   */
+  Model: Constructor<Model>;
+
+  /**
+   * A reference to Material constructor. Supports instanceof checks; this class
+   * is not directly constructable.
+   */
+  Material: Constructor<Material>;
+
+  /**
+   * A reference to PBRMetallicRoughness constructor. Supports instanceof
+   * checks; this class is not directly constructable.
+   */
+  PBRMetallicRoughness: Constructor<PBRMetallicRoughness>;
+}
+
+/**
+ * All events have a corresponding type string that can be used when adding
+ * listeners for them.
+ */
+export declare interface ThreeDOMEventMap {
+  'model-change': ModelChangeEvent;
+}
+
+/**
+ * A basic element in the 3DOM domain.
+ */
+export declare interface ThreeDOMElement {
+  /**
+   * A 3DOM element always has a reference to its Model of provenance unless it
+   * is the root of the scene graph (implictly the Model).
+   */
+  readonly ownerModel?: Model;
+}
+
+/**
+ * The ModelChangeEvent is dispatched globally whenever a model has loaded and
+ * been assigned to the global model property.
+ */
+export declare interface ModelChangeEvent extends Event {
+  /**
+   * A reference to the most recently assigned global model
+   */
+  model: Model;
+
+  /**
+   * A reference to the most recently replaced global model. Note that this
+   * model is no longer active and may no longer be mutated.
+   */
+  previousModel?: Model;
+}
+
+/**
+ * A Model is the root element of a 3DOM scene graph. It gives scripts access
+ * to the sub-elements found without the graph.
+ */
+export declare interface Model extends ThreeDOMElement {
+  /**
+   * An ordered set of unique Materials found in this model. The Materials are
+   * listed in scene graph traversal order.
+   */
+  readonly materials: Readonly<Material[]>;
+}
+
+/**
+ * A Material gives the script access to modify a single, unique material found
+ * in a model's scene graph.
+ *
+ * @see https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-material
+ */
+export declare interface Material extends ThreeDOMElement {
+  /**
+   * The name of the material, if any.
+   */
+  readonly name?: string;
+
+  /**
+   * The PBRMetallicRoughness configuration of the material.
+   */
+  readonly pbrMetallicRoughness: PBRMetallicRoughness;
```

Comment:
It's how it is organized in a glTF: https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#metallic-roughness-material

Reply:
I get that, but as you said earlier, JSON hierarchy does not necessarily equal ergonomic API shape. In this case I think it would be cleaner to follow something along three's organization here. I would rather use `instanceof` to determine what properties to expect.

Reply:
We are striving to make serialization back to glTF straight forward.

For decades, ergonomics on the web has been the domain of author code. That is what Three.js is, after all: a helper library that makes WebGL more ergonomic. Users will have a better sense of what ergonomic means for their use cases, and they will have a wide variety of opinions. Our goal is to facilitate them with capabilities.

Reply:
Fair enough; you have my opinion. You're right that it will work fine either way.

Reply:
👍 I agree with you that ergonomics suffer a lot for this decision. Let's chat more about this in internal channels.